From b2bd47c28cd197b9b894825297061c1a9691cd8c Mon Sep 17 00:00:00 2001
From: David du Colombier <djc@datadoghq.com>
Date: Fri, 17 Mar 2023 14:26:05 +0100
Subject: [PATCH 37/41] Free xmlDoc structure at the end of xccdf_session_load

When calling the xccdf_session_load_(xccdf|cpe|oval|tailoring)
functions, the session->source structure is recopied to
session->ds.session->component_sources, including the
source->xml.doc field, which contains the XML DOM.

This change adds a new oscap_source_free_xmlDoc function to the
OSCAP source API and call it at the end of the xccdf_session_load
function, so session->source->xml.doc is freed when it's not used
anymore.
---
 src/XCCDF/xccdf_session.c        |  4 +++-
 src/source/oscap_source.c        | 10 ++++++++++
 src/source/public/oscap_source.h |  6 ++++++
 3 files changed, 19 insertions(+), 1 deletion(-)

diff --git a/src/XCCDF/xccdf_session.c b/src/XCCDF/xccdf_session.c
index 54b2bd07a..e50af7997 100644
--- a/src/XCCDF/xccdf_session.c
+++ b/src/XCCDF/xccdf_session.c
@@ -716,7 +716,9 @@ int xccdf_session_load(struct xccdf_session *session)
 			return ret;
 		}
 	}
-	return xccdf_session_load_tailoring(session);
+	ret = xccdf_session_load_tailoring(session);
+	oscap_source_free_xmlDoc(session->source);
+	return ret;
 }
 
 static int _reporter(const char *file, int line, const char *msg, void *arg)
diff --git a/src/source/oscap_source.c b/src/source/oscap_source.c
index 879fc065c..c4f09ce30 100644
--- a/src/source/oscap_source.c
+++ b/src/source/oscap_source.c
@@ -153,6 +153,16 @@ void oscap_source_free(struct oscap_source *source)
 	}
 }
 
+void oscap_source_free_xmlDoc(struct oscap_source *source)
+{
+	if (source != NULL) {
+		if (source->xml.doc != NULL) {
+			xmlFreeDoc(source->xml.doc);
+			source->xml.doc = NULL;
+		}
+	}
+}
+
 /**
  * Returns human readable description of oscap_source origin
  */
diff --git a/src/source/public/oscap_source.h b/src/source/public/oscap_source.h
index 5c9f6472e..58f20baf5 100644
--- a/src/source/public/oscap_source.h
+++ b/src/source/public/oscap_source.h
@@ -95,6 +95,12 @@ OSCAP_API struct oscap_source *oscap_source_clone(struct oscap_source *old);
  */
 OSCAP_API void oscap_source_free(struct oscap_source *source);
 
+/**
+ * Dispose oscap_source xmlDoc structure.
+ * @param source Resource to dispose xmlDoc structure from
+ */
+OSCAP_API void oscap_source_free_xmlDoc(struct oscap_source *source);
+
 /**
  * Get filepath of the given resource
  * @memberof oscap_source
-- 
2.34.1

